readthedocs/vcs_support/backends/git.py

```diff
@@ -198,6 +194,24 @@ def clone(self):
 
         try:
             code, stdout, stderr = self.run(*cmd)
+
+            # TODO: for those VCS providers that don't tell us the `default_branch`
+            # of the repository in the incoming webhook,
+            # we need to get it from the cloned repository itself.
+            #
+            # cmd = ['git', 'symbolic-ref', 'refs/remotes/origin/HEAD']
+            # _, default_branch, _ = self.run(*cmd)
+            # default_branch = default_branch.replace('refs/remotes/origin/', '')
+            #
+            # The idea is to hit the APIv2 here to update the `latest` version with
+            # the `default_branch` we just got from the repository itself,
+            # after clonning it.
+            # However, we don't know the PK for the version we want to update.
```

Comment:
We could just add a `get_default_branch` method to the class here, and hit the API from the caller where we have the version?

https://github.com/readthedocs/readthedocs.org/blob/239d8905896d9bc9a7ec118b231a9c315f40f43c/readthedocs/doc_builder/director.py#L200-L204

Reply:
Hrm... We could, yes. I'm not sure if it's _better_ than just initializing this class with a _read_ `Version` object instead of the `version_slug` only.

* the `clone()` method's caller is the same class; from outside, the method called is `update()`
* I understand we would need to respect the calling order: we must call first `get_default_branch()` and then `update()`; otherwise, the default branch won't be defined

There are other potential solutions as well:

1. initialize the class with `version_pk` too
2. initialize the class with a complete `Version` object
3. use APIv3 (instead of APIv2) from the builder that supports getting a `Version` object _without the `pk`_: https://docs.readthedocs.io/en/stable/api/v3.html#version-detail


It probably makes sense to use APIv3 since it's part of the organic transition to remove APIv2 internally.

Reply:
Hrm, thinking a little more about this... These ides will interfere with this other improvement https://github.com/readthedocs/readthedocs.org/issues/9736 where we will be cloning the _exact_ branch directly, instead of the `default_branch` and then switching to the one wanted.

Reply:
This could also be part of another issue to investigate a little more, if you think it's not mandatory for this PR for now. It only affects Bitbucket currently.

Reply:
I think it's fine to not include, but I just wanted to note that we have options here to get the data we want by thinking about getting it another way. 🤷